File: LeetCode/89-gray-code/Solution3.java
import java.util.ArrayList;
import java.util.List;

/**
 * Generates the n-bit Gray Code sequence iteratively in a single loop.
 * The sequence ensures that consecutive values differ by exactly one bit.
 *
 * @param n The number of bits in the Gray Code sequence.
 * @return A list containing the Gray Code sequence in decimal form.
 */
class Solution3 {
  public List<Integer> grayCode(int n) {
    int sequenceLength = 1 << n; // Compute 2^n for the total sequence length
    List<Integer> result = new ArrayList<>(); // Initialize result list

    // Generate Gray Code sequence iteratively
    for (int i = 0; i < sequenceLength; i++) {
      result.add(i ^ (i >> 1)); // Compute Gray Code using XOR pattern
    }

    return result; // Return final sequence
  }
}

File: LeetCode/89-gray-code/solution3.ts
/**
 * Generates the n-bit Gray Code sequence iteratively in a single loop.
 * The sequence ensures that consecutive values differ by exactly one bit.
 *
 * @param {number} n - The number of bits in the Gray Code sequence.
 * @returns {number[]} An array containing the Gray Code sequence in decimal form.
 */
function grayCode(n: number): number[] {
  const sequenceLength = 1 << n; // Compute 2^n for the total length
  const result: number[] = []; // Initialize result array

  for (let i = 0; i < sequenceLength; i++) {
    result.push(i ^ (i >> 1)); // Generate Gray Code and store it
  }

  return result; // Return final sequence
}